Realm/Realm/Handles/SharedRealmHandle.cs

```diff
@@ -782,6 +796,17 @@ private static void HandleTaskCompletion<T>(IntPtr tcsPtr, Func<T> resultBuilder
             }
         }
 
+        private void OnTaskCancellation(uint? asyncTransactionHandle, TaskCompletionSource<object> tcs)
+        {
+            if (asyncTransactionHandle.HasValue)
+            {
+                NativeMethods.cancel_async_transaction(this, asyncTransactionHandle.Value, out var innerNativeException);
+                innerNativeException.ThrowIfNecessary();
```

Comment:
If a native exception occurs we should set it on the `tcs` rather than throw it here.